Refactoring, new view, bump up to 1.2

alexeyklyukin opened the pull request:
Add a view to show all client backends waiting on a lock for more than 5
minutes. The goal is to have a graph and a check reporting all long
client backend locks

Slight refactoring, use is_replica instead of pg_is_in_recovery()
consistently, use a saner way of checking Posgres versions, put
pgtelemetry-head in line with the latest version, use
client_stat_activity instead of pg_stat_activity in connections_by_ip
view, as the caller is clearly not interested in autovacuums and
friends.

Bump up to 1.2, update tests and migration files, put a version of the
expected file  for testing locally when pg_stat_statements is not in the
shared_preload_libraries.

mkabilov commented on extension/pgtelemetry--1.0--1.2.sql:
```diff
@@ -0,0 +1,55 @@
+-- version compatibility
+do $d$ begin
+if current_setting('server_version_num')::int < 100000 then
+   CREATE FUNCTION pg_current_xlog_location() RETURNS pg_lsn language sql as $$
```
should be different from the `-head` one?
in `-head` there's `CREATE OR REPLACE`

alexeyklyukin replied:
It should be create or replace here as well, thanks!

alexeyklyukin replied:
Removed the code altogether, dropping support for versions before Postgres 10.

alexeyklyukin commented:
Note that I've considered an approach of returning a query that would group locks by, say, each minute, so that we could be flexible in defining in grafana the time we wait before considering the not granted lock a problem; however, that would make the query slower and add some complexity without the clear win, so I opted to start with the simple hard-coded of 5 minutes for that.

mkabilov commented on extension/pgtelemetry--1.0--1.2.sql:
```diff
@@ -0,0 +1,55 @@
+
+create or replace view lock_queue_size_five_minutes_wait as
+select count(1) from pg_locks join client_stat_activity using(pid)
+where granted = 'f' and extract('epoch' from now() - query_start) > 300;
```
I wouldn't use view in this case at all; instead I'd put that query to the prometheus exporter or any other tool which will collect the metrics

mkabilov replied:
or maybe wrap that query into a function with the interval as an argument

alexeyklyukin replied:
Well, I was not sure what is the line between creating a view in the pgtelemetry and putting the bare query into the Postgres exporter, and I thought the view would make it easier to reuse the query (i.e. run it manually), as well as provide better visibility into the kind of queries we use for monitoring.

mkabilov replied:
changing of that interval value might be expensive, as it will lead to a new version of pgtelemetry

alexeyklyukin replied:
Right, I am on board with making it a function with the interval parameter, although I don't think we are going to change it that often (on the other hand, you never know... :-))

alexeyklyukin commented:
Some new changes, dropped support for PostgreSQL versions before 10 and removed the conditional code depending on the Postgres version. That resulted in some functions started using wal_ functions and some others pg_is_in_recovery() instead of is_replica(). Because of that, I had to re-create them in the migration script.

nick-adjust commented on extension/pgtelemetry--1.1--1.2.sql:
```diff
@@ -0,0 +1,75 @@
+drop function if exists pg_last_xlog_replay_location();
+
+create or replace view connections_by_ip_source as
+select client_addr, count(*) from client_stat_activity
```
maybe also `count(*) filter (where state != 'idle')` and the likes?

alexeyklyukin replied:
Not sure if we are not interested in idle connections here, one needs to check the callers of this view. The only reason I've included it in the changeset is that the original one counted non-backend connections as well (autovacuum, wal senders etc.), so I've excluded those.

nick-adjust replied:
I mean something like this:
```
select 
    client_addr, 
    count(*) as total_connections, 
    count(*) filter (where state = 'idle') as idle_connections,
    count(*) filter (where state = 'idle in transaction') as idle_in_tx_connections,
    ...
  from client_stat_activity
```

alexeyklyukin replied:
sadly this would break pg-management functions that rely on "count" being a column name in the output of this function; I propose to change it as a separate PR without bumping up the version of the extension.

nick-adjust commented:
aside from the comment above, lgtm
